internal_ws/ykshim/src/lib.rs

```diff
@@ -3,9 +3,124 @@
 //! For more information, see this section in the documentation:
 //! https://softdevteam.github.io/ykdocs/tech/yk_structure.html
 
-/// The production API.
-mod prod_api;
-
-/// The testing API.
-/// These functions are only exposed to allow testing from the external workspace.
 mod test_api;
+
+use std::ffi::{c_void, CString};
+use std::os::raw::c_char;
+use ykcompile::CompiledTrace;
+use yktrace::{
+    sir::{SirTrace, SIR},
+    tir::TirTrace,
+    ThreadTracer,
+};
+
+/// The different ways by which we can collect a trace.
+#[derive(Clone, Copy)]
+#[repr(u8)]
+#[allow(dead_code)]
+enum TracingKind {
+    /// Software tracing via ykrustc.
+    SoftwareTracing = 0,
+    /// Hardware tracing via ykrustc + hwtracer.
+    HardwareTracing = 1,
+}
+
+/// Starts tracing using the specified kind, returning a ThreadTracer through which to stop
+/// tracing.
+#[no_mangle]
+unsafe extern "C" fn __ykshim_start_tracing(tracing_kind: u8) -> *mut ThreadTracer {
+    let tracing_kind = match tracing_kind {
+        0 => yktrace::TracingKind::SoftwareTracing,
+        1 => yktrace::TracingKind::HardwareTracing,
+        _ => return std::ptr::null_mut(),
+    };
+    let tracer = yktrace::start_tracing(tracing_kind);
+    Box::into_raw(Box::new(tracer))
+}
+
+/// Stops tracing, consuming the ThreadTracer and returning an opaque pointer to a SirTrace. If an
+/// error occurs then the returned pointer will be NULL and `error_msg` will contain details of the
+/// error.
+#[no_mangle]
+unsafe extern "C" fn __ykshim_stop_tracing(
+    tracer: *mut ThreadTracer,
+    error_msg: *mut *mut c_char,
+) -> *mut SirTrace {
+    let tracer = Box::from_raw(tracer);
+    let sir_trace = tracer.stop_tracing();
+    match sir_trace {
+        Ok(sir_trace) => Box::into_raw(Box::new(sir_trace)),
+        Err(err) => {
+            *error_msg = CString::new(err.to_string())
+                .unwrap_or_else(|err| {
```

Comment:
I'm not sure if all this code does anything meaningfully useful over just `unwrap`.

Reply:
The code attempts to give a meaningful error message as to what's gone wrong in the (opaque from Rust's POV) C code. But I think the `unwrap_or_else` might be wrong. Wouldn't it need to be `and_then()`?

If we just let this panic, then we don't make use of the error message the C code has constructed.

Reply:
(and we can't use `panic!` on the message since it's not constant)

Reply:
sorry, i misread my own code. Yes, we can panic here.